services/brig/src/Brig/Data/Client.hs

```diff
@@ -187,19 +188,44 @@ updatePrekeys u c pks = do
         _ -> return False
 
 claimPrekey :: UserId -> ClientId -> AppIO (Maybe ClientPrekey)
-claimPrekey u c = withOptLock u c $ do
-  prekey <- retry x1 $ query1 userPrekeys (params Quorum (u, c))
-  case prekey of
-    Just (i, k) -> do
-      if i /= lastPrekeyId
-        then retry x1 $ write removePrekey (params Quorum (u, c, i))
-        else
-          Log.debug $
-            field "user" (toByteString u)
-              . field "client" (toByteString c)
-              . msg (val "last resort prekey used")
-      return $ Just (ClientPrekey c (Prekey i k))
-    Nothing -> return Nothing
+claimPrekey u c = do
+  randomStrategy <- view randomPrekeys 
+  -- Log.warn $ field "randomStrategy" $ show randomStrategy
+  case randomStrategy of
+    -- Use DynamoDB based optimistic locking strategy
+    Nothing -> withOptLock u c $ do
+      prekey <- retry x1 $ query1 userPrekey (params Quorum (u, c))
+      removeAndReturnPreKey prekey
+    -- Use random prekey selection strategy
+    Just () -> do
+      lock <- view randomPrekeyLocalLock 
+      withLocalLock lock $ do
+        prekeys <- retry x1 $ query userPrekeys (params Quorum (u, c))
+        prekey <- pickRandomPrekey prekeys
+        removeAndReturnPreKey prekey
+
+  where
+    removeAndReturnPreKey :: Maybe (PrekeyId, Text) -> AppIO (Maybe ClientPrekey)
+    removeAndReturnPreKey (Just (i, k)) = do
+          if i /= lastPrekeyId
+            then retry x1 $ write removePrekey (params Quorum (u, c, i))
+            else
+              Log.debug $
+                field "user" (toByteString u)
+                  . field "client" (toByteString c)
+                  . msg (val "last resort prekey used")
+          return $ Just (ClientPrekey c (Prekey i k))
+    removeAndReturnPreKey Nothing = return Nothing
+
+    pickRandomPrekey :: [(PrekeyId, Text)] -> AppIO (Maybe (PrekeyId, Text))
+    pickRandomPrekey [] = return Nothing 
+    -- unless we only have one key left
+    pickRandomPrekey [pk] = return $ Just pk
+    -- pick among list of keys, except lastPrekeyId
+    pickRandomPrekey pks = do
```

Comment:
We should not actually pick the last resort prekey unless all other prekeys are used (it gives us less security because it will be re-used). So instead I would suggest to pick the random prekey from all but the last one and only pick the last resort key if there is no key left.

Reply:
Hey @franziskuskiefer that's exactly what this code does unless I misunderstand 😄 
 - if we have only one key left we return that one, [pattern matching the singleton list on line 223](https://github.com/wireapp/wire-server/pull/1416/files#diff-4902a73060613f1a4272a1912557acc1698ef6bf17271e658e3d6f7f067a65edR223) (it should be the `lastPrekey` but we don't actually check that) and the [`removeAndReturnPreKey`](https://github.com/wireapp/wire-server/pull/1416/files#diff-4902a73060613f1a4272a1912557acc1698ef6bf17271e658e3d6f7f067a65edR208) won't delete it from the database.
 - if we have more than one key left (pattern matching any other list that's neither empty nor singleton):
   - we filter out the `lastPrekeyId` from the list of all prekeys [into a `pks'` list on line 226](https://github.com/wireapp/wire-server/pull/1416/files#diff-4902a73060613f1a4272a1912557acc1698ef6bf17271e658e3d6f7f067a65edR226)
   - and we pick at random in the filtered out list `pks'`.  ([line 227](https://github.com/wireapp/wire-server/pull/1416/files#diff-4902a73060613f1a4272a1912557acc1698ef6bf17271e658e3d6f7f067a65edR227))

So the `lastPrekeyId` never get used until it's the last one remaining.

Reply:
Ah right. I didn't properly look at line 226, i.e that `pks'` doesn't contain the last resort prekey. Then this is doing the right thing! 🙂